src/testing/slim_check/sampleable.lean

```diff
@@ -196,18 +257,108 @@ sampleable_char 3 " 0123abcABC:,;`\\/"
 
 variables {α}
 
-/-- implementation of `sampleable (list α)` -/
-def list.shrink' (shrink_a : α → lazy_list α) : list α → lazy_list (list α)
+section list_shrink
+
+variables [has_sizeof α] (shr : Π x : α, lazy_list { y // sizeof_lt y x })
+
+lemma list.sizeof_drop_lt_sizeof_of_lt_length {xs : list α} {k}
+  (hk : 0 < k) (hk' : k < xs.length) :
+  sizeof (list.drop k xs) < sizeof xs :=
+begin
+  induction xs generalizing k,
+  { cases hk' },
+  cases k,
+  { cases hk },
+  have : sizeof xs_tl < sizeof (xs_hd :: xs_tl),
+  { unfold_wf, linarith },
+  cases k,
+  { simp [this] },
+  { simp,
+    transitivity,
+    { solve_by_elim [xs_ih, lt_of_succ_lt_succ hk', zero_lt_succ] },
+    { assumption } }
+end
+
+lemma list.sizeof_cons_lt_right (a b : α) {xs : list α} (h : sizeof a < sizeof b) :
+  sizeof (a :: xs) < sizeof (b :: xs) :=
+by unfold_wf; assumption
+
+lemma list.sizeof_cons_lt_left (x : α) {xs xs' : list α} (h : sizeof xs < sizeof xs') :
+  sizeof (x :: xs) < sizeof (x :: xs') :=
+by unfold_wf; assumption
+
+lemma list.sizeof_append_lt_left {xs ys ys' : list α} (h : sizeof ys < sizeof ys') :
+  sizeof (xs ++ ys) < sizeof (xs ++ ys') :=
+begin
+  induction xs,
+  { apply h },
+  { unfold_wf, simp [list.sizeof], exact xs_ih }
+end
+
+lemma list.one_le_sizeof (xs : list α) : 1 ≤ sizeof xs :=
+by cases xs; unfold_wf; [refl, linarith]
+
+/--
+`list.shrink_removes` shrinks a list by removing chunks of size `k` in
+the middle of the list.
+-/
+def list.shrink_removes (k : ℕ) (hk : 0 < k) : Π (xs : list α) n, n = xs.length → lazy_list { ys // sizeof_lt ys xs }
+| xs n hn :=
+  if hkn : k > n then lazy_list.nil
+  else
+  if hkn' : k = n then
+    have 1 < xs.sizeof,
+      by { subst_vars, cases xs, { contradiction },
+           unfold_wf, apply lt_of_lt_of_le,
+           show 1 < 1 + has_sizeof.sizeof xs_hd + 1, { linarith },
+           { mono, apply list.one_le_sizeof, } },
+    lazy_list.singleton ⟨[], this ⟩
+  else
+    have h₂ : k < xs.length, from hn ▸ lt_of_le_of_ne (le_of_not_gt hkn) hkn',
+    match list.split_at k xs, rfl : Π ys, ys = list.split_at k xs → _ with
+    |  ⟨xs₁,xs₂⟩, h :=
+      have h₄ : xs₁ = xs.take k,
+        by simp [list.split_at_eq_take_drop] at h; tauto,
+      have h₃ : xs₂ = xs.drop k,
+        by simp [list.split_at_eq_take_drop] at h; tauto,
+      have sizeof xs₂ < sizeof xs,
+        by rw h₃; solve_by_elim [list.sizeof_drop_lt_sizeof_of_lt_length],
+      have h₁ : n - k = xs₂.length,
+        by simp [h₃, ← hn],
+      have h₅ : ∀ (a : list α), sizeof_lt a xs₂ → sizeof_lt (xs₁ ++ a) xs, from
+        λ a h, by rw [← list.take_append_drop k xs, ← h₃, ← h₄]; solve_by_elim [list.sizeof_append_lt_left],
+      lazy_list.cons ⟨xs₂, this⟩ $ subtype.map ((++) xs₁) h₅ <$> list.shrink_removes xs₂ (n - k) h₁
```

Comment:
This is a well-founded recursive call, right? And you're only using this function for evaluation? Can you confirm that it actually evaluates with reasonable performance? The equation compiler sometimes produces horrible byte code for non-meta well founded recursive calls, even when the same function tagged as `meta` makes completely reasonable byte code. It might be better to make this function `meta` even though it doesn't have to be. Or is too much of `slim_check` in non-meta land?

Reply:
Same thing applies elsewhere, e.g. for trees

Reply:
I haven't noticed any performance problems due to that. Things got faster with this change. I can keep the suggestion in mind and apply it when it becomes necessary.

A lot of `slim_check` is on purpose in non-`meta` land. Proofs play a role in making sure instances are well-behaved. They also play a role because the termination of `minimize` is compromised if some `shrink` definitions are not well thought out.

Reply:
For reference, @gebner pointed this out to me [here](https://github.com/leanprover-community/mathlib/pull/3004#discussion_r439755022). I looked at the trace as he recommended at the time, and indeed he was right, the meta version was much cleaner. So this is worth keeping an eye on.